### DISCUSSION TITLE
Quick Start Tutorial is not working - An error occurred The requested redirect URL is not permitted.

### DISCUSSION THREAD

**Discussion by dinhbang** [etsy/open-api#1217] [Show and tell]
Hello,

I have do connect API with https://developers.etsy.com/documentation/tutorials/quickstart But, It is not working. 
My server.js file below:

`// Import the express and fetch libraries
const express = require('express');
const fetch = require("node-fetch");
const hbs = require("hbs");

// Create a new express application
const app = express();
app.set("view engine", "hbs");
app.set("views", `${process.cwd()}/views`);

// Send a JSON response to a default get request
app.get('/ping', async (req, res) => {
	const requestOptions = {
		'method': 'GET',
		'headers': {
			'x-api-key': 'iibu5bcmzso6pXXXXXXX'',
		},
	};

	const response = await fetch(
		'https://api.etsy.com/v3/application/openapi-ping',
		requestOptions
	);

	if (response.ok) {
		const data = await response.json();
		res.send(data);
	} else {
		res.send("oops");
	}
});

// This renders our `index.hbs` file.
app.get('/', async (req, res) => {
	res.render("index");
});

/**
These variables contain your API Key, the state sent
in the initial authorization request, and the client verifier compliment
to the code_challenge sent with the initial authorization request
 */
const clientID = 'iibu5bcmzso6pXXXXXXX';
const clientVerifier = 'kWjT3PdpzuldTgbRuiiVpBgRWh0p4RSDUNzb0FjfjQE';
const redirectUri = 'http://localhost:3003/oauth/redirect';

app.get("/oauth/redirect", async (req, res) => {
	// The req.query object has the query params that Etsy authentication sends
	// to this route. The authorization code is in the `code` param
	const authCode = req.query.code;
	const tokenUrl = 'https://api.etsy.com/v3/public/oauth/token';
	const requestOptions = {
		method: 'POST',
		body: JSON.stringify({
			grant_type: 'authorization_code',
			client_id: clientID,
			redirect_uri: redirectUri,
			code: authCode,
			code_verifier: clientVerifier,
		}),
		headers: {
			'Content-Type': 'application/json'
		}
	};

	// Extract the access token from the response access_token data field
	if (response.ok) {
		const tokenData = await response.json();
		res.redirect(`/welcome?access_token=${tokenData.access_token}`);
	} else {
		res.send("oops");
	}
});

app.get("/welcome", async (req, res) => {
	// We passed the access token in via the querystring
	const { access_token } = req.query;

	// An Etsy access token includes your shop/user ID
	// as a token prefix, so we can extract that too
	const user_id = access_token.split('.')[0];

	const requestOptions = {
		headers: {
			'x-api-key': clientID,
			// Scoped endpoints require a bearer token
			Authorization: `Bearer ${access_token}`,
		}
	};

	const response = await fetch(
		`https://api.etsy.com/v3/application/users/${user_id}`,
		requestOptions
	);

	if (response.ok) {
		const userData = await response.json();
		// Load the template with the first name as a template variable.
		res.render("welcome", {
			first_name: userData.first_name
		});
	} else {
		res.send("oops");
	}
});

// Start the server on port 3003
const port = 3003;
app.listen(port, () => {
	console.log(`Example app listening at http://localhost:${port}`);
});`
It can be show ping screen below:
![image_2024_03_13T08_45_14_713Z](https://github.com/etsy/open-api/assets/6812151/af5f877d-8bb4-4bc8-bc34-ae07a8349729)

The oauth URL is:  

> https://www.etsy.com/oauth/connect?response_type=code&redirect_uri=http://localhost:3003/oauth/redirect&scope=email_r&client_id=iibu5bcmzso6pXXXXXXX&state=0otd59&code_challenge=7P7iQmvHeKdIEBojI-wHiNNsUhUyNm6a-O82l211osI&code_challenge_method=S256

When I click to "Authenticate with Etsy". It is show error "

> An error occurred
>The requested redirect URL is not permitted.

**Comment by EnsarBatuhanAyd** [comment]
ı am taking exactly to same error with same process :/ any help?

**Comment by eschramm** [comment]
It could be that the redirect URI doesn't match the Callback URLs configured for your app and/or it requires HTTPS given it's OAuth.

**Comment by slider1234** [comment]
i posted this thread a while back to simplify the horrid tutorial. Try it out at https://github.com/etsy/open-api/discussions/1208
Also, i agree with the last comment saying, "It could be that the redirect URI doesn't match the Callback URL". I made this same mistake at one point. The Etsy Developer tutorial just made it much more complicated than it needed to be.

**Comment by sdumper** [comment]
i have nev er had this problem in the past but i have a new authorized store as of 11/01/2024 and am getting the same bullshit response! so pissed.  An error occurred
The requested redirect URL is not permitted.

note im not a noob at this i have 6 active stores! wtf etsy!!!

**Comment by Wilderbill** [comment]
I strongly recommend use if the Python script provide by slider1234 has noted above.

But, even with that I struggled for days with the inability to authorize certain scope combinations.
The cause of that problem was multiple overlapping instances as revealed by "Apps You Use".
( See the website: https://www.etsy.com/your/apps?ref=si_apps )
I had a half dozen active instances of my app as a result of my flailing attempts to get it working.
( Apparently, there can be multiple instances of your app with overlapping, conflicting scopes. )
The solution was to delete all of my instances and start again.
Set the scopes you want and run: https://github.com/slider1234/ETSY-API-Auto-Authenticator again.
( This does not require a repeat of the 3-week application process. )

**Reply by slider1234** [reply]
This is the updated version 
https://github.com/etsy/open-api/discussions/1236

it looks to be the same content I have at the GitHub  link you shared as well

**Comment by Crazyglue** [comment]
Still getting this issue. Following the quick start exactly, getting the error occurred issue. Triple checked that the callback url was correct, but still nothing. Not sure why the official docs dont work. This API V3 transition has been really really rocky for me

**Comment by jumpdumpling** [comment]
If you also need Ai assistance specific to your codebase, Grok does very well with the Etsy API. It will walk you through an updated quickstart tutorial based on your tech stack.
